clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -109,6 +125,8 @@ public ProduceRequest(Struct struct, short version) {
      */
     @Override
     public Struct toStruct() {
+        // Store it in a local variable to protect against concurrent updates
+        Map<TopicPartition, MemoryRecords> partitionRecords = partitionRecordsOrFail();
```

Comment:
@junrao, I added that to make the code fail-fast in case assumptions are broken in the future. Calling `toStruct` after `clearPartitionRecords` is a bug and we should fail with an appropriate error message if that happens.